blocks/search.py

```diff
@@ -0,0 +1,401 @@
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+from six.moves import range
+
+import numpy
+
+from theano import config
+from theano import function
+from theano import tensor
+
+from blocks.filter import VariableFilter
+from blocks.graph import ComputationGraph
+
+floatX = config.floatX
+
+
+class Search(object):
+    """Abstract search class.
+
+    Parameters
+    ----------
+    sequence_generator : sequence generator
+        Sequence generator to be used
+
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, sequence_generator):
+        self.generator = sequence_generator
+        self.compiled = False
+
+    @abstractmethod
+    def compile(self):
+        self.compiled = True
+
+    @abstractmethod
+    def search(self, **kwargs):
+        r"""Performs search.
+
+        Parameters
+        ----------
+        \*\*kwargs :
+            Arguments needed by sequence generator
+
+        Returns
+        -------
+        Generated sequences
+
+        """
+        if not self.compiled:
+            self.compile()
+
+
+def unchunk_rename(*args, **kwargs):
+    def _trace(func):
+        def wrapper(self, *inputs):
+            merged_input = []
+            for input in inputs:
+                merged_input += [OrderedDict([(name, self.merge_chunks(value))
+                                              for name, value
+                                              in input.iteritems()])]
+            names, outputs = func(self, *merged_input)
+            if add_time_dim:
+                outputs = [self.divide_by_chunks(value.reshape((1,) +
+                                                               value.shape))
+                           for value in outputs]
+            else:
+                outputs = [self.divide_by_chunks(value.reshape(value.shape))
+                           for value in outputs]
+            return OrderedDict(zip(names, outputs))
+        return wrapper
+    if len(args) == 1 and callable(args[0]):
+        # No arguments, this is the decorator
+        # Set default values for the arguments
+        add_time_dim = True
+        return _trace(args[0])
+    else:
+        # This is just returning the decorator
+        add_time_dim = kwargs['add_time_dim']
+        return _trace
+
+
+class BeamSearch(Search):
+    """Beam search.
+
+    Parameters
+    ----------
+    beam_size : int
+        Size of beam.
+    batch_size : int
```

Comment:
I am not sure the beam search needs batch size. This makes the code significantly more complicated. When beam size small it should be fast on its own and when beam size is large that would give you the required degree of parallelization that you need in order to benefit from GPU.